blog/_posts/2025-04-04-Optimize-AKS-Traffic-with-externalTrafficPolicy=Local.md

```diff
@@ -0,0 +1,157 @@
+
+- **Health Probe on Each Node:** Azure automatically configures a health probe on the load balancer that targets the `HealthCheckNodePort` across all nodes in the LB's backend pool. Kubernetes ensures that this port only returns a successful response on nodes that are running at least one *ready* pod for the Service. Nodes with no pods for that Service will fail the health check.
+
+- **Load Balancer Backend Pool:** With `externalTrafficPolicy=Local`, all cluster nodes are listed in the LB's backend pool. But due to the health probes, nodes without a Service pod are marked **unhealthy** and won't receive traffic​. Only nodes with healthy pods respond to the probe and remain in rotation. By contrast, in `Cluster` mode, every node responds (since even if it has no pod, kube-proxy will forward the traffic), so the LB sees all nodes as healthy​. The `kube-proxy` component manages this port and ensures that it only responds as healthy if the node hosts at least one healthy pod for the service.
+
+ **IPTables Rules**: IPTables rules are configured to only forward incoming traffic from the Azure Load Balancer (ALB) directly to pods running on the same node. These rules ensure that traffic is never forwarded to other nodes. This localized traffic routing reduces latency and ensures that external connections continue to be served even during node update operations.
+
+By combining these mechanisms, `externalTrafficPolicy=Local` provides a robust way to manage external traffic while maintaining source IP visibility and ensuring traffic is routed to healthy pods only.
+
+## Best Practices to gracefully close existing connections and shut service pods
+
+Gracefully handling pod shutdowns is critical to maintaining service reliability and avoiding disruptions, especially in scenarios involving HTTP keep-alive connections or long-lived client sessions. Without a graceful shutdown process, external customers could see errors like - `connection refused` and `connection reset by peer` during node related events.
+
+Below are detailed best practices for how pods can handle Kubernetes initiated termination requests (eg: pod evictions or a scale down) to ensure a smooth shutdown process.
+
+### Gracefully Closing Existing Connections
+
+When a pod is shutting down (receiving the `TERM` signal), it is essential to ensure that existing client connections are closed properly to avoid abrupt disconnections or errors. Failing to handle this gracefully could result in clients encountering errors like `connection reset by peer` or `connection refused`, leading to a poor user experience and potential service disruptions.
+
+- **For HTTP/1.1 Connections**:
+    The server should include a `Connection: close` header in its response for all active and new incoming requests. This informs clients not to reuse the connection and allows idle connections to be closed gracefully.
```

Comment:
I am not sure whether you should recommend this as a general rule.  Please do some research about `TIME_WAIT`.  
Also, for TLS 1.3, golang has a known bug that does not handle close_notify properly which cause RST packets.